File: Battleship/src/main/java/de/htwg/battleship/actor/childactor/grandchild/CollisionActor.java
// CollisionController.java

package de.htwg.battleship.actor.childactor.grandchild;

import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.UntypedActor;
import de.htwg.battleship.actor.messages.CollisionMessage;
import de.htwg.battleship.model.IShip;
import de.htwg.battleship.util.StatCollection;

/**
 * Controller for checking if ship collides. use the Chain-of-Responsibility-Pattern
 *
 * @author Moritz Sauter (SauterMoritz@gmx.de)
 * @version 1.00
 * @since 2014-12-04
 */
public class CollisionActor extends UntypedActor {

    private final ActorRef same =
        getContext().actorOf(Props.create(SameCollision.class), "both_same");
    private final ActorRef different = getContext()
        .actorOf(Props.create(DifferentCollision.class), "first_true");

    @Override
    public void onReceive(Object message) throws Exception {
        if (!(message instanceof CollisionMessage)) {
            unhandled(message);
            return;
        }
        CollisionMessage collisionMessage = (CollisionMessage) message;
        if (collisionMessage.getShip1().isOrientation() ==
            collisionMessage.getShip2().isOrientation()) {
            // orientations are the same
            same.forward(message, getContext());
        } else {
            // orientations are different
            different.forward(message, getContext());
        }
    }
}

abstract class ShipCollisionActor extends UntypedActor {
    @Override
    public void onReceive(Object message) throws Exception {
        if (!(message instanceof CollisionMessage)) {
            unhandled(message);
            return;
        }
        final boolean collision =
            handleCollisionMessage((CollisionMessage) message);
        getSender().tell(collision, getSelf());
    }

    abstract boolean handleCollisionMessage(final CollisionMessage message);
}

class SameCollision extends ShipCollisionActor {
    private boolean isCollision(final IShip firstShip, final IShip secondShip) {
        final int firstLow, firstFix, secLow, secFix;
        if (firstShip.isOrientation()) {
            firstLow = firstShip.getX();
            firstFix = firstShip.getY();
            secLow = secondShip.getX();
            secFix = secondShip.getY();
        } else {
            firstLow = firstShip.getY();
            firstFix = firstShip.getX();
            secLow = secondShip.getY();
            secFix = secondShip.getX();
        }
        final int firstUp = firstLow + firstShip.getSize() - 1;
        final int secUp = secLow + secondShip.getSize() - 1;
        for (int i = secLow; i <= secUp; i++) {
            if (StatCollection.isBetween(firstUp, firstLow, i) &&
                secFix == firstFix) {
                return true;
            }
        }
        return false;
    }

    @Override
    boolean handleCollisionMessage(final CollisionMessage message) {
        return isCollision(message.getShip1(), message.getShip2());
    }
}

class DifferentCollision extends ShipCollisionActor {
    /**
     * Check if the ships collide. To use this method the ships must have
     * different orientations.
     *
     * @param vertical   the first ship
     * @param horizontal the second ship with a different orientation
     *
     * @return true if the ships collide
     */
    private boolean checkShipsDifferent(IShip vertical, IShip horizontal) {
        final int horizontalLow = horizontal.getX();
        final int horizontalUp = horizontalLow + horizontal.getSize() - 1;
        final int horizontalFix = horizontal.getY();
        final int verticalLow = vertical.getY();
        final int verticalUp = verticalLow + vertical.getSize() - 1;
        final int verticalFix = vertical.getX();
        return
            StatCollection.isBetween(verticalUp, verticalLow, horizontalFix) &&
            StatCollection.isBetween(horizontalUp, horizontalLow, verticalFix);
    }

    @Override
    boolean handleCollisionMessage(final CollisionMessage message) {
        IShip vertical, horizontal;
        if (message.getShip1().isOrientation()) {
            horizontal = message.getShip1();
            vertical = message.getShip2();
        } else {
            horizontal = message.getShip2();
            vertical = message.getShip1();
        }
        return checkShipsDifferent(vertical, horizontal);
    }
}

File: Battleship/src/main/scala/de/htwg/battleship/actor/childs/CollisionActor.scala
package de.htwg.battleship.actor.childs

import akka.actor.{Actor, Props}
import akka.util.Timeout
import de.htwg.battleship.actor.messages.CollisionMessage
import de.htwg.battleship.model.IShip
import de.htwg.battleship.util.StatCollection

import scala.concurrent.duration._
import scala.util.control.Breaks._


object CollisionActor {
    val ACTOR_NAME = "collider"
}

class CollisionActor extends Actor {
    val same = context.actorOf(Props[SameCollision], SameCollision.ACTOR_NAME)
    val diff = context.actorOf(Props[DifferentCollision], DifferentCollision.ACTOR_NAME)
    implicit val timeout = Timeout(5 seconds)

    /**
      * Method to check if there is a collision.
      *
      * @return Sends 'true' for it is all okay or 'false' if there is a problem to the asker
      */
    override def receive: Receive = {
        case msg: CollisionMessage =>
            if (msg.ship1.isOrientation == msg.ship2.isOrientation) {
                same forward msg
            } else {
                diff forward msg
            }
        case msg => unhandled(msg)
    }
}

object SameCollision {
    val ACTOR_NAME = "collider_same"
}

class SameCollision extends Actor {
    override def receive: Receive = {
        case msg: CollisionMessage =>
            // negate result as we need false as stronger boolean in the composition
            val result = !isCollision(msg.ship1, msg.ship2)
            sender() ! result
    }

    /**
      * Help method to check if two ships collide when both ships have the same orientation
      *
      * @param firstShip  first ship
      * @param secondShip second ship
      * @return true if there was a collision, false if everything was okay
      */
    def isCollision(firstShip: IShip, secondShip: IShip): Boolean = {
        var firstLow: Int = -1
        var firstFix: Int = -1
        var secLow: Int = -1
        var secFix: Int = -1
        if (firstShip.isOrientation) {
            firstLow = firstShip.getX
            firstFix = firstShip.getY
            secLow = secondShip.getX
            secFix = secondShip.getY
        } else {
            firstLow = firstShip.getY
            firstFix = firstShip.getX
            secLow = secondShip.getY
            secFix = secondShip.getX
        }
        val firstUp = firstLow + firstShip.getSize - 1
        val secUp = secLow + secondShip.getSize - 1
        if (secFix != firstFix) {
            false
        } else {
            var returner = false
            breakable {
                // loop INCLUDING secLow AND secUp -> for (i = secLow; i <= secUp; i++)
                for (i <- secLow to secUp) {
                    if (StatCollection.isBetween(firstUp, firstLow, i)) {
                        returner = true
                        break
                    }
                }
            }
            returner
        }
    }
}

object DifferentCollision {
    val ACTOR_NAME = "collider_different"
}

class DifferentCollision extends Actor {
    override def receive: Receive = {
        case msg: CollisionMessage =>
            var horizontal: IShip = msg.ship1
            var vertical: IShip = msg.ship2
            if (!msg.ship1.isOrientation) {
                horizontal = msg.ship2
                vertical = msg.ship1
            }
            // negate result as we need false forces the entire result to be false
            val result = !checkShipsDifferent(vertical, horizontal)
            sender() ! result
    }

    /**
      * Help method to check if two ships collide if they have different orientations.
      *
      * @param vertical   the ship with orientation = false
      * @param horizontal the ship with the orientation = true
      * @return 'true' if there is a collision or false if everything is okay
      */
    def checkShipsDifferent(vertical: IShip, horizontal: IShip): Boolean = {
        val horizontalLow = horizontal.getX
        val horizontalUp = horizontalLow + horizontal.getSize - 1
        val horizontalFix = horizontal.getY
        val verticalLow = vertical.getY
        val verticalUp = verticalLow + vertical.getSize - 1
        val verticalFix = vertical.getX
        StatCollection.isBetween(verticalUp, verticalLow, horizontalFix) &&
            StatCollection.isBetween(horizontalUp, horizontalLow, verticalFix)
    }
}

